lib/galaxy/webapps/galaxy/api/workflows.py

```diff
@@ -991,6 +996,276 @@ def _generate_report(self, trans, invocation_id, **kwd):
+            'contributors': contributors,
+            'license': 'https://spdx.org/licenses/CC-BY-4.0.html',
+        }
+
+        keywords = []
+        for tag in stored_workflow.tags:
+            keywords.append(tag.user_tname)
+        for tag in history.tags:
+            if tag.user_tname not in keywords:
+                keywords.append(tag.user_tname)
+
+        metrics = {}
+        tools, input_subdomain, output_subdomain, pipeline_steps, software_prerequisites = [], [], [], [], []
+        for i, step in enumerate(workflow_invocation.steps):
+            if step.workflow_step.type == 'tool':
+                workflow_outputs_list , output_list, input_list = [], [], []
+                for wo in step.workflow_step.workflow_outputs:
+                    workflow_outputs_list.append(wo.output_name)
+                for job in step.jobs:
+                    metrics[i] = summarize_job_metrics(trans, job)
+                    for job_input in job.input_datasets:
+                        if hasattr(job_input.dataset, 'dataset_id'):
+                            encoded_dataset_id = trans.security.encode_id(job_input.dataset.dataset_id)
+                            input_obj = {
+                                'filename': job_input.dataset.name,
+                                'uri': url_for('history_content', history_id=encoded_history_id, id=encoded_dataset_id, qualified=True),
+                                'access_time': job_input.dataset.create_time.isoformat(),
+                            }
+                            input_list.append(input_obj)
+
+                    for job_output in job.output_datasets:
+                        if hasattr(job_output.dataset, 'dataset_id'):
+                            encoded_dataset_id = trans.security.encode_id(job_output.dataset.dataset_id)
+                            output_obj = {
+                                'filename': job_output.dataset.name,
+                                'uri': url_for('history_content', history_id=encoded_history_id, id=encoded_dataset_id, qualified=True),
+                                'access_time': job_output.dataset.create_time.isoformat(),
+                            }
+                            output_list.append(output_obj)
+
+                        if job_output.name in workflow_outputs_list:
+                            output = {
+                                'mediatype': job_output.dataset.extension,
+                                'uri': {
+                                    'filename': job_output.dataset.name,
+                                    'uri': url_for('history_content', history_id=encoded_history_id, id=encoded_dataset_id, qualified=True),
+                                    'access_time': job_output.dataset.create_time.isoformat(),
+                                }
+                            }
+                            output_subdomain.append(output)
+                current_tool = dict_workflow['steps'][str(i)]
```

Comment:
That is a workflow step, not a tool.

Reply:
Renamed in https://github.com/galaxyproject/galaxy/pull/10217/files